cpp/src/column/column_view.cpp

```diff
@@ -0,0 +1,123 @@
+/*
+ * Copyright (c) 2019, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <cudf/column/column_view.hpp>
+#include <cudf/null_mask.hpp>
+#include <cudf/types.hpp>
+#include <utilities/error_utils.hpp>
+
+#include <exception>
+#include <vector>
+
+namespace cudf {
+
+namespace detail {
+column_view_base::column_view_base(data_type type, size_type size,
+                                   void const* data,
+                                   bitmask_type const* null_mask,
+                                   size_type null_count, size_type offset)
+    : _type{type},
+      _size{size},
+      _data{data},
+      _null_mask{null_mask},
+      _null_count{null_count},
+      _offset{offset} {
+  CUDF_EXPECTS(size >= 0, "Column size cannot be negative.");
+
+  if (type.id() == EMPTY) {
+    _null_count = size;
+    CUDF_EXPECTS(nullptr == data, "EMPTY column should have no data.");
+    CUDF_EXPECTS(nullptr == null_mask,
+                 "EMPTY column should have no null mask.");
+  } else if (size > 0) {
+    CUDF_EXPECTS(nullptr != data, "Null data pointer.");
+  }
+
+  CUDF_EXPECTS(offset >= 0, "Invalid offset.");
+
+  if ((null_count > 0) and (type.id() != EMPTY)) {
+    CUDF_EXPECTS(nullptr != null_mask,
+                 "Invalid null mask for non-zero null count.");
+  }
+}
+
+// If null count is known, returns it. Else, compute and return it
+size_type column_view_base::null_count() const {
+  if (_null_count > cudf::UNKNOWN_NULL_COUNT) {
+    return _null_count;
+  } else {
+    _null_count = cudf::count_unset_bits(null_mask(), offset(), size());
+    return null_count();
```

Comment:
Why not return `_null_count` here rather than the tail-recursive call?

Reply:
+1. I would write this function as:

```
size_type column_view_base::null_count() const {
  if (_null_count == cudf::UNKNOWN_NULL_COUNT) {
    _null_count = cudf::count_unset_bits(null_mask(), offset(), size());
  }
  return _null_count;
}
```